Remora.Discord.Extensions/Services/PermissionComputationService.cs

```diff
@@ -0,0 +1,198 @@
+//
+//  PermissionComputationService.cs
+//
+//  Author:
+//       Jarl Gullberg <jarl.gullberg@gmail.com>
+//
+//  Copyright (c) Jarl Gullberg
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Lesser General Public License for more details.
+//
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Remora.Discord.API.Abstractions.Objects;
+using Remora.Discord.API.Abstractions.Rest;
+using Remora.Discord.API.Objects;
+using Remora.Discord.Commands.Extensions;
+using Remora.Discord.Commands.Services;
+using Remora.Rest.Core;
+using Remora.Results;
+
+namespace Remora.Discord.Extensions.Services;
+
+/// <inheritdoc />
+public class PermissionComputationService : IPermissionComputationService
+{
+    private readonly ContextInjectionService _contextInjection;
+    private readonly IDiscordRestGuildAPI _guildAPI;
+    private readonly IDiscordRestChannelAPI _channelAPI;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="PermissionComputationService"/> class.
+    /// </summary>
+    /// <param name="contextInjection">The context injection service.</param>
+    /// <param name="guildAPI">The guild API.</param>
+    /// <param name="channelAPI">The channel API.</param>
+    public PermissionComputationService
+    (
+        ContextInjectionService contextInjection,
+        IDiscordRestGuildAPI guildAPI,
+        IDiscordRestChannelAPI channelAPI
+    )
+    {
+        _contextInjection = contextInjection;
+        _guildAPI = guildAPI;
+        _channelAPI = channelAPI;
+    }
+
+    /// <inheritdoc/>
+    public async Task<Result<IDiscordPermissionSet>> ComputeMemberPermissions
+    (
+        Snowflake userID,
+        Snowflake guildID,
+        Snowflake? channelID = null,
+        CancellationToken ct = default
+    )
+    {
+        var getMember = await _guildAPI.GetGuildMemberAsync(guildID, userID, ct);
+        if (!getMember.IsDefined(out var member))
+        {
+            return Result<IDiscordPermissionSet>.FromError(getMember);
```

Comment:
Doesn't `FromError(Result)` wrap? Curious because this could potentially be a little problematic depending on how the result is handled (e.g. being thrown up the call stack, where there's already issues with result wrapping as-is)

Reply:
It does wrap, but there have been changes made to how wrapping actually functions so it shouldn't be as much of an issue as it previously was. `FromError` uses the error set in the passed result directly, so you should always have the nearest concrete error object on the result object you get.

I think you may be thinking of RestError and its related problems, which have some other considerations (but should be far more readable and useful in the latest release of Remora.Rest).

On top of that, FromError is required since the result types differ :P

Reply:
I was actually thinking of Remora.Discord.Commands and the issue of four-deep wrapping

Reply:
Furthermore doesn't `FromError` accept an `IResultError` that returns a failed result instead of a wrapped failed result?

Reply:
There are several overloads (of which one does take `T where T : IResultError`), but in this case it's using this overload <https://github.com/Remora/Remora.Results/blob/main/Remora.Results/Result.cs#L148>, which doesn't cause any nesting in the previously problematic way.